realm/realm-library/src/main/cpp/io_realm_internal_OsObject.cpp

```diff
@@ -0,0 +1,205 @@
+    ObjectWrapper& operator=(ObjectWrapper&&) = delete;
+
+    ObjectWrapper(ObjectWrapper const&) = delete;
+    ObjectWrapper& operator=(ObjectWrapper const&) = delete;
+
+    ~ObjectWrapper()
+    {
+    }
+};
+
+struct ChangeCallback {
+    ChangeCallback(ObjectWrapper* wrapper)
+        : m_wrapper(wrapper)
+    {
+    }
+
+    void parse_fields(JNIEnv* env, CollectionChangeSet const& change_set)
+    {
+        if (m_field_names_array) {
+            return;
+        }
+
+        if (!change_set.deletions.empty()) {
+            m_deleted = true;
+            return;
+        }
+
+        std::vector<jstring> field_names;
+        auto table = m_wrapper->m_object.row().get_table();
+        for (size_t i = 0; i < change_set.columns.size(); ++i) {
+            if (change_set.columns[i].empty()) {
+                continue;
+            }
+            // FIXME: After full integration of the OS schema, parse the column name from
+            // wrapper->m_object.get_object_schema() will be faster.
+            field_names.push_back(to_jstring(env, table->get_column_name(i)));
+        }
+        m_field_names_array = env->NewObjectArray(field_names.size(), java_lang_string, 0);
+        for (size_t i = 0; i < field_names.size(); ++i) {
+            env->SetObjectArrayElement(m_field_names_array, i, field_names[i]);
+        }
+    }
+
+    JNIEnv* check_env()
+    {
+        JNIEnv* env = JniUtils::get_env(false);
+        if (!env || env->ExceptionCheck()) {
+            // JVM dettached or java exception has been thrown before.
+            return nullptr;
+        }
+        return env;
+    }
+
+    void before(CollectionChangeSet const& change_set)
+    {
+        JNIEnv* env = check_env();
+        if (!env) {
+            return;
+        }
+
+        parse_fields(env, change_set);
+    }
+
+    void after(CollectionChangeSet const& change_set)
+    {
+        JNIEnv* env = check_env();
+        if (!env) {
+            return;
+        }
+        if (change_set.empty()) {
+            return;
+        }
+
+        parse_fields(env, change_set);
+
+        m_wrapper->m_row_object_weak_ref.call_with_local_ref(env, [&](JNIEnv*, jobject row_obj) {
+            static JavaMethod notify_change_listeners(env, row_obj, "notifyChangeListeners",
+                                                      "([Ljava/lang/String;)V");
+            env->CallVoidMethod(row_obj, notify_change_listeners, m_field_names_array);
+        });
+    }
+
+    void error(std::exception_ptr err)
+    {
+        if (err) {
+            try {
+                std::rethrow_exception(err);
+            }
+            catch (const std::exception& e) {
+                Log::e("Caught exception in object change callback %1", e.what());
+            }
+        }
+    }
+
+private:
+    ObjectWrapper* m_wrapper;
+    bool m_deleted = false;
+    jobjectArray m_field_names_array = nullptr;
+};
+
+static void finalize_object(jlong ptr);
```

Comment:
forward declaration ... i can remove it . Build failed without it before if I remembered correctly... but it builds now.